SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialRPCService.cpp

```diff
@@ -0,0 +1,430 @@
+		const ServerEndpoint* Endpoint = View->GetComponentData<ServerEndpoint>(EntityId);
+		LastAckedRPCIds.Add(EntityRPCType(EntityId, ERPCType::ServerReliable), Endpoint->ReliableRPCAck);
+		LastAckedRPCIds.Add(EntityRPCType(EntityId, ERPCType::ServerUnreliable), Endpoint->UnreliableRPCAck);
+		LastSentRPCIds.Add(EntityRPCType(EntityId, ERPCType::ClientReliable), Endpoint->ReliableRPCBuffer.LastSentRPCId);
+		LastSentRPCIds.Add(EntityRPCType(EntityId, ERPCType::ClientUnreliable), Endpoint->UnreliableRPCBuffer.LastSentRPCId);
+		break;
+	}
+	case SpatialConstants::MULTICAST_RPCS_COMPONENT_ID:
+	{
+		const MulticastRPCs* Component = View->GetComponentData<MulticastRPCs>(EntityId);
+		LastSentRPCIds.Add(EntityRPCType(EntityId, ERPCType::NetMulticast), Component->MulticastRPCBuffer.LastSentRPCId);
+
+		// Duplicate the initial multicast RPCs?
+		break;
+	}
+	default:
+		checkNoEntry();
+	}
+}
+
+void SpatialRPCService::OnEndpointAuthorityLost(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
+{
+	switch (ComponentId)
+	{
+	case SpatialConstants::CLIENT_ENDPOINT_COMPONENT_ID:
+	{
+		LastAckedRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ClientReliable));
+		LastAckedRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ClientUnreliable));
+		LastSentRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ServerReliable));
+		LastSentRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ServerUnreliable));
+		ClearOverflowedRPCs(EntityId);
+		break;
+	}
+	case SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID:
+	{
+		LastAckedRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ServerReliable));
+		LastAckedRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ServerUnreliable));
+		LastSentRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ClientReliable));
+		LastSentRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ClientUnreliable));
+		ClearOverflowedRPCs(EntityId);
+		break;
+	}
+	case SpatialConstants::MULTICAST_RPCS_COMPONENT_ID:
+	{
+		LastSentRPCIds.Remove(EntityRPCType(EntityId, ERPCType::NetMulticast));
+		break;
+	}
+	default:
+		checkNoEntry();
+	}
+}
+
+void SpatialRPCService::ExtractRPCsForType(Worker_EntityId EntityId, ERPCType Type)
+{
+	uint64 LastSeenRPCId;
+	EntityRPCType EntityTypePair = EntityRPCType(EntityId, Type);
+
+	if (Type == ERPCType::NetMulticast)
+	{
+		LastSeenRPCId = LastSeenMulticastRPCIds[EntityId];
+	}
+	else
+	{
+		LastSeenRPCId = LastAckedRPCIds[EntityTypePair];
+	}
+
+	const RPCRingBuffer& Buffer = GetBufferFromView(EntityId, Type);
+
+	RPCRingBufferDescriptor Descriptor = RPCRingBufferUtils::GetRingBufferDescriptor(Type);
+
+	if (Buffer.LastSentRPCId >= LastSeenRPCId)
+	{
+		for (uint64 RPCId = LastSeenRPCId + 1; RPCId <= Buffer.LastSentRPCId; RPCId++)
+		{
+			uint32 ElementIndex = Descriptor.GetRingBufferElementIndex(RPCId);
+			if (Buffer.RingBuffer[ElementIndex].IsSet())
+			{
+				bool bKeepExtracting = ExtractRPCCallback.Execute(EntityId, Type, Buffer.RingBuffer[ElementIndex].GetValue());
+				if (!bKeepExtracting)
+				{
+					break;
```

Comment:
Need to make sure we only ack the RPCs that we have processed. So the last ID executed -1.